### PR TITLE
[FEATURE] Automatically rev component versions on Github Actions PR #6218

### PR DESCRIPTION
**Pull request by lcaresia**
### <samp>🤖 Generated by Copilot at 8ab554c</samp>

This pull request improves the `git-diff-on-components` action to automatically update component versions when dependencies change. It also updates the licenses of the action dependencies and the version of the `new-request-created` component that depends on a modified `common.mjs` file.

### <samp>🤖 Generated by Copilot at 8ab554c</samp>

> _We are the masters of the code, we change the versions as we please_
> _We update the licenses, we respect the dependencies_
> _We add the functions and the logic, we format the files with a newline_
> _We are the git-diff-on-components, we make the components shine_

### <samp>🤖 Generated by Copilot at 8ab554c</samp>

*  Add functions to extract and update component versions ([link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-96a10ab03c7ca3d08adbc6e17341ada96d47039f58e9fac8afecaa18a5ad2ff7R98-R105))
*  Set git user and email to pull request author ([link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-96a10ab03c7ca3d08adbc6e17341ada96d47039f58e9fac8afecaa18a5ad2ff7L339-R372))
*  Loop over components with modified dependencies and update their versions using `sed` ([link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-96a10ab03c7ca3d08adbc6e17341ada96d47039f58e9fac8afecaa18a5ad2ff7L339-R372))
*  Commit and push changes to pull request branch with message "Automatically updating actions and sources versions" ([link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-96a10ab03c7ca3d08adbc6e17341ada96d47039f58e9fac8afecaa18a5ad2ff7L339-R372))
*  Update version of `new-request-created` component to `0.0.2` in `new-request-created.mjs` ([link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-091740d58fb24ea491b8ddba645cb0ebcb066e93dd2f2cf262fddcb88d4a6e60L6-R6))
*  Remove license text of `@vercel/ncc` package from `licenses.txt` ([link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-44edc623550c297902993c4b4c6b6246b950b0b630f39e6e4b32c6475c4d31fbL240-L249))
*  Add license texts of `buffer-from`, `has-flag`, `source-map-support`, and `supports-color` packages to `licenses.txt` ([link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-44edc623550c297902993c4b4c6b6246b950b0b630f39e6e4b32c6475c4d31fbR356-R380), [link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-44edc623550c297902993c4b4c6b6246b950b0b630f39e6e4b32c6475c4d31fbR1261-R1273), [link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-44edc623550c297902993c4b4c6b6246b950b0b630f39e6e4b32c6475c4d31fbR2263-R2287), [link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-44edc623550c297902993c4b4c6b6246b950b0b630f39e6e4b32c6475c4d31fbR2342-R2354))
*  Add newline at the end of `index.js` ([link](https://github.com/PipedreamHQ/pipedream/pull/6295/files?diff=unified&w=0#diff-96a10ab03c7ca3d08adbc6e17341ada96d47039f58e9fac8afecaa18a5ad2ff7L355-L354))

### CODE REVIEW

File: .github/actions/git-diff-on-components/src/index.js [hunk 1]
```diff
@@ -95,6 +95,14 @@ function includesVersion(contents) {
   return contents.match(new RegExp(/version: "\d+\.\d+\.\d+"/g)) || contents.match(new RegExp(/"version": "\d+\.\d+\.\d+"/g));
 }
 
+function getVersion(contents) {
+  return includesVersion(contents)[0].slice(-6, -1)
+}
+
+function increaseVersion(version) {
+  return version.slice(0, 4) + (+version.split('.')[2] + 1)
```
Comment (andrewjschuang) [inline .github/actions/git-diff-on-components/src/index.js:103]: Same question as above

File: .github/actions/git-diff-on-components/src/index.js [hunk 2]
```diff
@@ -336,20 +344,35 @@ async function run() {
     componentsDiffContents = await checkVersionModification(componentsPendingForGitDiff, componentsThatDidNotModifyVersion);
   }
 
-  const totalErrors = componentsThatDidNotModifyVersion.length + componentsDiffContents.length;
+  const totalErrors = componentsThatDidNotModifyVersion.length;
   let counter = 1;
 
+  await execCmd("git", ["config", "--global", "user.email", "lucascarezia@gmail.com"]);
+  await execCmd("git", ["config", "--global", "user.name", "Lucas Caresia"]);
```
Comment (andrewjschuang) [inline .github/actions/git-diff-on-components/src/index.js:351]: Are these lines needed for the `git commit`?
If so, we better use Pipedream data, like
```
email: pipedream@pipedream.com
name: Pipedream
```

File: .github/actions/git-diff-on-components/src/index.js [hunk 3]
```diff
@@ -336,20 +344,35 @@ async function run() {
...
-  componentsDiffContents.forEach(({ dependencyFilePath, componentFilePath }) => {
-    console.log(`${counter++}) You need to change the version of ${getComponentFilePath(componentFilePath)} since dependency file ${getComponentFilePath(dependencyFilePath)} was modified.`);
-  });
+  if (componentsDiffContents.length) {
+    for ({ dependencyFilePath, componentFilePath } of componentsDiffContents) {
+      const content = await readFile(componentFilePath, "utf-8")
+      const currentVersion = getVersion(content)
+      const increasedVersion = increaseVersion(currentVersion)
+
+      await execCmd("sed", ["-i", `0,/${currentVersion}/{s/${currentVersion}/${increasedVersion}/}`, getComponentFilePath(componentFilePath)]);
```
Comment (andrewjschuang) [inline .github/actions/git-diff-on-components/src/index.js:363]: For QA: We need to make sure this command doesn't change any other line besides the component version

File: .github/actions/git-diff-on-components/src/index.js [hunk 4]
```diff
@@ -336,20 +344,35 @@ async function run() {
...
+
+      console.log(`✅ Version of ${getComponentFilePath(componentFilePath)} changed from ${currentVersion} to ${increasedVersion} since dependency file ${getComponentFilePath(dependencyFilePath)} was modified.`);
+    };
+
+    await execCmd("git", ["add", "."]);
+    await execCmd("git", ["commit", "-m", "Automatically updating actions ans sources versions"]);
+    await execCmd("git", ["push", "--force-with-lease", "--no-verify"]);
```
Comment (andrewjschuang) [inline .github/actions/git-diff-on-components/src/index.js:370]: Looks like there is an existing [`git-auto-commit-action`](https://github.com/stefanzweifel/git-auto-commit-action). Can you please check this out?

**Review by andrewjschuang — CHANGES_REQUESTED** [review]
Thanks @lcaresia, can you please take a look at the comments?

File: .github/actions/git-diff-on-components/src/index.js (hunk 2 above)
Comment (lcaresia) [reply .github/actions/git-diff-on-components/src/index.js:351]: Yes, unfortunately is necessary.

File: .github/actions/git-diff-on-components/src/index.js [hunk 5]
```diff
@@ -336,20 +347,35 @@ async function run() {
...
+      const currentVersion = getVersion(content)
+      const increasedVersion = increaseVersion(currentVersion)
+
+      await execCmd("sed", ["-i", `0,/${currentVersion}/{s/${currentVersion}/${increasedVersion}/}`, getComponentFilePath(componentFilePath)]);
+
+      console.log(`✅ Version of ${getComponentFilePath(componentFilePath)} changed from ${currentVersion} to ${increasedVersion} since dependency file ${getComponentFilePath(dependencyFilePath)} was modified.`);
+    };
+
+    await execCmd("git", ["add", "."]);
+    await execCmd("git", ["commit", "-m", "Automatically updating actions ans sources versions"]);
```
Comment (GTFalcao) [inline .github/actions/git-diff-on-components/src/index.js:372]:
```suggestion
    await execCmd("git", ["commit", "-m", "Automatically updating actions and sources versions"]);
```

**Review by GTFalcao — COMMENTED** [review]
Hi @lcaresia , thanks for the contribution!

I left one comment on a typo.

I also added to one of @andrewjschuang 's previous comments.

File: .github/actions/git-diff-on-components/src/index.js [hunk 6]
```diff
@@ -95,6 +95,14 @@ function includesVersion(contents) {
   return contents.match(new RegExp(/version: "\d+\.\d+\.\d+"/g)) || contents.match(new RegExp(/"version": "\d+\.\d+\.\d+"/g));
 }
 
+function getVersion(contents) {
+  return includesVersion(contents)[0].slice(-6, -1)
```
[the remark this answers is not in the document]
Comment (GTFalcao) [reply .github/actions/git-diff-on-components/src/index.js]: This was changed to `.match(/\d{3}\.\d{3}\.\d{3}/)[0]` but that looks for exactly 3 digits (in each number) - shouldn't this be any amount of digits instead?
e.g.
`(\d+\.){2}\d+`
which would match any version numbers in the x.y.z format

File: .github/actions/git-diff-on-components/src/index.js [hunk 7]
```diff
@@ -95,6 +95,17 @@ function includesVersion(contents) {
   return contents.match(new RegExp(/version: "\d+\.\d+\.\d+"/g)) || contents.match(new RegExp(/"version": "\d+\.\d+\.\d+"/g));
 }
 
+function getVersion(contents) {
+  return includesVersion(contents)[0].match(/\d{3}\.\d{3}\.\d{3}/)[0];
```
Comment (GTFalcao) [inline .github/actions/git-diff-on-components/src/index.js:99]: Can you check this regexp as mentioned in [this previous conversation](https://github.com/PipedreamHQ/pipedream/pull/6295#discussion_r1186457476)?

The way it is now matches exactly 3 digits for each version number, which I believe is not the desired behavior.

**Review by GTFalcao — CHANGES_REQUESTED** [review]
Hi @lcaresia , please check this one remaining comment, I had added to a previous comment by Andrew and linked it to the review this time.

File: .github/actions/git-diff-on-components/src/index.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lcaresia) [reply .github/actions/git-diff-on-components/src/index.js]: You are completely corect, I didn't see your comment.

**Review by GTFalcao — APPROVED** [review]
Looks good to me!

**Review by andrewjschuang — CHANGES_REQUESTED** [review]
Hi @lcaresia, I'm QA'ing using [these 2 tests](https://github.com/PipedreamHQ/pipedream/issues/6218#issuecomment-1543850277) in [this PR](https://github.com/PipedreamHQ/pipedream/pull/6404).

**Test 1 has failed:**
`TypeError: Cannot read properties of null (reading '0')`


image

**Skipped Test 2 since Test 1 failed.**

File: components/accelo/accelo.app.mjs [hunk 8]
```diff
@@ -13,7 +13,7 @@ export default {
 
         return companies.map((company) => ({
           value: company.id,
-          label: company.name,
+          label: company.name, // just testing, this comment need be removed after
```
Comment (dylburger) [inline components/accelo/accelo.app.mjs:16]:
```suggestion
          label: company.name, // Just testing, remove this comment later.
```

Comment (dylburger) [reply components/accelo/accelo.app.mjs:16]: Changed to sentence case and clarified instruction according to Google's Developer docs style guide.

File: .github/workflows/pull-request-checks.yaml [hunk 9]
```diff
@@ -39,6 +39,13 @@ jobs:
           all_files: ${{ steps.changed_files.outputs.all }}
           base_commit: ${{ github.event.pull_request.base.sha }}
           head_commit: ${{ github.event.pull_request.head.sha }}
+      - uses: stefanzweifel/git-auto-commit-action@v4
+        with:
+           commit_message: Automatically updating actions and sources versions
```
Comment (dylburger) [inline .github/workflows/pull-request-checks.yaml:44]:
```suggestion
commit_message: Automatically updating action and source versions
```

File: .github/workflows/pull-request-checks.yaml [hunk 10]
```diff
@@ -39,6 +39,13 @@ jobs:
...
+           commit_options:  '--no-verify --signoff'
+           commit_user_name: Pipedream [Bot]
```
[the remark this answers is not in the document]
Comment (dylburger) [reply .github/workflows/pull-request-checks.yaml:46]: Removed unnecessary double quotes

File: .github/workflows/pull-request-checks.yaml [hunk 11]
```diff
@@ -39,6 +39,13 @@ jobs:
...
+           commit_user_email: pipedream@pipedream.com
```
Comment (dylburger) [inline .github/workflows/pull-request-checks.yaml:47]:
```suggestion
commit_user_email: pipedream@pipedream.com
```

Comment (dylburger) [reply .github/workflows/pull-request-checks.yaml:47]: Removed unnecessary double quotes

File: .github/workflows/pull-request-checks.yaml [hunk 12]
```diff
@@ -39,6 +39,13 @@ jobs:
...
+           commit_author: Pipedream [Bot] <pipedream@pipedream.com>
```
Comment (dylburger) [inline .github/workflows/pull-request-checks.yaml:48]:
```suggestion
commit_author: Pipedream [Bot] <pipedream@pipedream.com>
```

Comment (dylburger) [reply .github/workflows/pull-request-checks.yaml:48]: Removed unnecessary double quotes

**Comment by omar721588951** [comment]
[1 (1) (1) (1) (2) (2) (1) (1) (1) (1).pdf](https://github.com/PipedreamHQ/pipedream/files/11548297/1.1.1.1.2.2.1.1.1.1.pdf)
extrasion suave

**Review by andrewjschuang — DISMISSED** [review]
Thanks for doing this! LGTM!

**Review by andrewjschuang — APPROVED** [review]

**Pull request merged**